containers/base/Dockerfile

```diff
@@ -1,252 +1,293 @@
+# earlier layer).
+    /tools/node/bin/npm install \
+        ws@1.1.4 \
+        http-proxy@1.13.2 \
+        mkdirp@0.5.1 \
+        node-uuid@1.4.7 \
+        bunyan@1.7.1 \
+        tcp-port-used@0.1.2 \
+        node-cache@3.2.0 && \
+    cd / && \
+    /tools/node/bin/npm install -g forever && \
+
+# Clean up
+    apt-get autoremove -y && \
+    rm -rf /var/lib/apt/lists/* && \
+    rm -rf /tmp/* && \
+    rm -rf /root/.cache/* && \
+    rm -rf /usr/share/locale/* && \
+    rm -rf /usr/share/i18n/locales/* && \
+    cd /
+
+ENV LC_ALL en_US.UTF-8
+ENV LANG en_US.UTF-8
+
+ADD config/ipython.py /etc/ipython/ipython_config.py
+ADD config/nbconvert.py /etc/jupyter/jupyter_notebook_config.py
+
+# Directory "py" may be empty and in that case it will git clone pydatalab from repo
+ADD pydatalab /datalab/lib/pydatalab
+ADD nbconvert /datalab/nbconvert
+
+# Do IPython configuration and install build artifacts
+# Then link stuff needed for nbconvert to a location where Jinja will find it.
+# I'd prefer to just use absolute path in Jinja imports but those don't work.
+RUN ipython profile create default && \
+    jupyter notebook --generate-config
+
+RUN if [ -d /datalab/lib/pydatalab/.git ]; then \
+        echo "use local lib"; \
+      else \
+        git clone https://github.com/googledatalab/pydatalab.git /datalab/lib/pydatalab; \
+      fi
+RUN cd /datalab/lib/pydatalab && \
+    /tools/node/bin/npm install -g typescript && \
+    tsc --module amd --noImplicitAny --outdir datalab/notebook/static datalab/notebook/static/*.ts && \
+    tsc --module amd --noImplicitAny --outdir google/datalab/notebook/static google/datalab/notebook/static/*.ts && \
+    /tools/node/bin/npm uninstall -g typescript
+RUN cd /datalab/lib/pydatalab && \
+    pip install --upgrade-strategy only-if-needed --no-cache-dir . && \
+    pip install --upgrade-strategy only-if-needed /datalab/lib/pydatalab/solutionbox/image_classification/. && \
+    pip install --upgrade-strategy only-if-needed /datalab/lib/pydatalab/solutionbox/structured_data/. && \
+    pip3 install --upgrade-strategy only-if-needed --no-cache-dir . && \
+    pip3 install --upgrade-strategy only-if-needed /datalab/lib/pydatalab/solutionbox/image_classification/. && \
+    pip3 install --upgrade-strategy only-if-needed /datalab/lib/pydatalab/solutionbox/structured_data/. && \
+    jupyter nbextension install --py datalab.notebook && \
+    jupyter nbextension install --py google.datalab.notebook && \
+    jupyter nbextension enable --py widgetsnbextension && \
+    rm datalab/notebook/static/*.js google/datalab/notebook/static/*.js && \
+    mkdir -p /datalab/nbconvert && \
+    cp -R /usr/local/share/jupyter/nbextensions/gcpdatalab/* /datalab/nbconvert && \
+    ln -s /usr/local/lib/python2.7/dist-packages/notebook/static/custom/custom.css /datalab/nbconvert/custom.css && \
+    cd /
+
+## Use Debian unstable via pinning -- new style via APT::Default-Release

```

Comment:
What is the impact of this? Does this affect any subsequent apt-get installs in the container, or that the user might do? The "unstable" bit makes me wonder... is there a way to avoid this?